### DISCUSSION TITLE
v6.0.0-pre.2

### DISCUSSION THREAD

**Discussion by coolsoftwaretyler** [mobxjs/mobx-state-tree#2164] [General] [answered]
## Breaking Changes

* Bump typescript from 3.9.10 to 5.3.3 by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2146
* Eliminate `NonEmptyObject` by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2152
* Improved typing for union types by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2151
* fix #1530 array clear replace should produce one patch by @BrianHung in https://github.com/mobxjs/mobx-state-tree/pull/2073
* Feat/has env and getenv changes by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2163

## Features

* We added a new feature, `hasEnv`, to support https://github.com/mobxjs/mobx-state-tree/pull/2163

## Fixes

* Union types will now be fixed, but this also may break TypeScript for some users
* Array operations produce condensed patches when we can, but this may break for some users who relied on the patch generation as it was

## Tests

No new test-only contributions

## Docs

* Docs/update docs homepage by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2141
* docs: fix typos in docs update by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2142
* Docs: add comparison between React Context/Reducer by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2158

## Community/Developer changes

* Bump typescript from 3.9.10 to 5.3.3 by @thegedge in https://github.com/mobxjs/mobx-state-tree/pull/2146
* Chore/use bun as a package manager and script runner by @coolsoftwaretyler in https://github.com/mobxjs/mobx-state-tree/pull/2148

## New Contributors
* @thegedge made their first contribution in https://github.com/mobxjs/mobx-state-tree/pull/2146

**Full Changelog**: https://github.com/mobxjs/mobx-state-tree/compare/v5.4.1...v6.0.0-pre.2

<hr /><em>This discussion was created from the release <a href='https://github.com/mobxjs/mobx-state-tree/releases/tag/v6.0.0-pre.2'>v6.0.0-pre.2</a>.</em>

**Comment by coolsoftwaretyler** [comment]
Here's an attempt at a migration guide from v5 to v6:

## TypeScript update

Make sure you're using TypeScript `5.3.3`, and you have the following compiler flags:

```json
{
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "noImplicitAny": true,
    "noImplicitReturns": true,
    "noImplicitThis": true
}
```

Or the shorthand:

```json
{
  "strict": true,
  "noImplicitReturns": true
}
```

## Removing `$nonEmptyObject`

We removed the `$nonEmptyObject` key from the `ModelCreationType`. For the most part, this was an internal feature that would only have shown up in TypeScript errors. There is no runtime change, but if you have any custom typings that relied on something like `[$nonEmptyObject]` - you should be able to eliminate those.

## Union Type Changes

In the past, named enumerations had more specific type than unnamed enumerations. You may have some TypeScript errors where you were using anonymous enumerations. You can type those with the union of literal values inside the enumeration now.

```ts
import { t } from "mobx-state-tree";

/**
 * In MobX-State-Tree 5.4.1, this is typed as:
 * ISimpleType<"Red" | "Orange" | "Green">
 */
const namedEnum = t.enumeration("Color", ["Red", "Orange", "Green"]);

/**
 * In MobX-State-Tree 5.4.1, this is typed as:
 * ISimpleType<string>
 */
const anonymousEnum = t.enumeration(["Red", "Orange", "Green"]);

/**
 * If you use mobx-state-tree@^6.0.0, both of these will be typed as:
 * ISimpleType<"Red" | "Orange" | "Green">
 */
```

This has also fixed https://github.com/mobxjs/mobx-state-tree/issues/1525 and https://github.com/mobxjs/mobx-state-tree/issues/1664

For #1525, If you had models with properties that were unions including other models, you may have [done some type assertions](https://github.com/mobxjs/mobx-state-tree/issues/1525#issuecomment-1614885456) to resolve the issue. This should no longer be necessary. In most cases, we don't expect this to show up as errors. But the inferred types *have* changed. Look through your codebase for unions of models in the properties of other models, and if you hit any issues, you may be able to use better typing form here on out.

The issue presented itself in #1664 when using types with different creation and instance types which were passed to unions inside of arrays. If you see more TypeScript errors aside from the other items mentioned, look for `types.union` that creates a union which includes `Array`, `Map`, or `Model` types. If you had any custom type assertions, you may no longer need those.

## Array operation patch changes

When an `Array` type calls `clear`, or splices itself from index `0`, the generated patch is different. We have shortened the operations to be a patch like:

```
op: "replace", path: "", value: []
```

If you use `onPatch` and expect any specific shape of patches from `Array.clear` or `Array.splice`, you may need to update your app logic to handle the changed patch generation.

## getEnv now throws

If you call `getEnv` and there is no environment, it will now throw an error. Previously, it would return an empty object.

Check your codebase for `getEnv` calls, and guard against this error with `hasEnv`, which will return `true` if the current state tree has an environment, or `false` if not.

**Reply by thegedge** [reply]
<3 that looks great! I'm excited for that patch improvement.

**Comment by coolsoftwaretyler** [comment, marked as answer]
I pulled a few fixes in from `5.4.2` into a `6.0.0-pre.3` release and published. We've got a maintainer meeting in a few weeks, I will propose that we mark v6 as latest at that time. I'll send a message in the discord, and move the migration notes here over to the other discussion.
